ledger/participant-integration-api/src/main/scala/platform/indexer/JdbcIndexer.scala

```diff
@@ -335,39 +352,52 @@ private[indexer] class JdbcIndexer private[indexer] (
       case CommandRejected(recordTime, submitterInfo, reason) =>
         ledgerDao.storeRejection(Some(submitterInfo), recordTime.toInstant, offset, reason)
     }
-    result.map(_ => ())(DEC)
+    result.map(_ => ())(DirectExecutionContext)
   }
 
   private class SubscriptionResourceOwner(
       readService: ReadService,
   )(implicit loggingContext: LoggingContext)
       extends ResourceOwner[IndexFeedHandle] {
     override def acquire()(implicit context: ResourceContext): Resource[IndexFeedHandle] =
-      Resource(Future {
-        val (killSwitch, completionFuture) = readService
-          .stateUpdates(startExclusive)
-          .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
-          .mapAsync(1) {
-            case (offset, update) =>
-              withEnrichedLoggingContext(JdbcIndexer.loggingContextFor(offset, update)) {
-                implicit loggingContext =>
-                  Timed.future(
-                    metrics.daml.indexer.stateUpdateProcessing,
-                    handleStateUpdate(offset, update),
-                  )
-              }
-          }
-          .toMat(Sink.ignore)(Keep.both)
-          .run()
-
-        new SubscriptionIndexFeedHandle(killSwitch, completionFuture.map(_ => ()))
-      })(
+      Resource(Future(handleUpdates()))(
         handle =>
           for {
             _ <- Future(handle.killSwitch.shutdown())
             _ <- handle.completed.recover { case NonFatal(_) => () }
           } yield ()
       )
+
+    private def handleUpdates(): SubscriptionIndexFeedHandle = {
+      import com.daml.dec.DirectExecutionContext
+
+      val (killSwitch, completionFuture) = readService
+        .stateUpdates(startExclusive)
+        .viaMat(KillSwitches.single)(Keep.right[NotUsed, UniqueKillSwitch])
+        .mapAsync(concurrency) {
```

Comment:
I think this has consistency problems:
1. Party `A` sends a command to Participant `P1` that creates contract `C1`. Party `B`  is observer on contract `C1`, but is hosted on Participant `P2`.
1. Participant `P2` (for whatever reason) is slow in processing updates or has had a downtime and hasn't seen the creation of `C1` yet.
1. Party `A` sends another command that archives `C1` again.
1. Participant `P2` now catches up and both the create and archive of `C1` in the same `concurrency` batch and processes these in parallel. If the archiving is processed before the creation, we'll end up with an active contract in `participant_contracts` that could be used by command interpretations, and a create event not marked as archived in `participant_events.create_consumed_at`, making `C1` show up in the active contracts.

Reply:
I think you're right. Would it be worth tweaking the index queries so that they use the ledger end as a cap, or will this still cause issues?

Reply:
I think concurrent updates is a good idea though! But we still need something sequential which catches up with all those updates, and maintains consistency.
Being able to set up the indexer this way would allow us to scale horizontally certain DB operations as well! But I think it would be also quite tricky to implement -> therefore it will be explored first where is the ceiling for sequential updates/batched updates

What you could do here is to put every preparation, on a mapAsync, and sequential sql execution on the next stage (probably defeats the purpose)

And what is actually already on the horizon and considered as a low hanging fruit: do expensive computation on a mapAsync (basically the stage before the db operation as far as I understood)

Reply:
> I think you're right. Would it be worth tweaking the index queries so that they use the ledger end as a cap, or will this still cause issues?

That is another issue, but that doesn't solve non-deterministic processing of creates and archives.

> What you could do here is to put every preparation, on a mapAsync, and sequential sql execution on the next stage (probably defeats the purpose)

Yes, you probably want to add some pipelining so that `fetching from the ledger store` > `converting to state updates` > `preparing index inserts and serialization` > `actual DB operation` are done with async boundaries and buffers between each stage so that faster stages can at least already prepare the work for downstream.

Reply:
yeah, sounds exactly how I perceived!

Reply:
At the very least, we could parallelize preparing the SQL, and run it in sequence.

We could also re-orient `DbDispatcher` so it just holds a `Connection` and works off a queue of SQL operations, which should help.

Reply:
> At the very least, we could parallelize preparing the SQL, and run it in sequence.

I believe that running the preparation fully in parallel is not going to be particularly effective, because it's much faster than execution it and it will just produce a growing backlog of transactions to persist without increasing overall throughput.

However, pipelining the preparation and the execution can be interesting, mostly because at the moment the preparation also deserializes the transaction from the `ReadService` and serializes the DAML-LF values to be persisted. The preparation is therefore than the execution part (ratio of times I've seen is around 1:5), but still costly enough for us to get a speed up from pipelining the two operations.

If we switch to a zero-copy approach for objects in our system, this could however end up negating the gain we could get from pipelining.

Reply:
Ooops, sorry, just noticed that was already said multiple times in the thread above, apologies.

Nice to have a very solid agreement on this though. 🙂